include/mw/consensus/KernelSumValidator.h

```diff
@@ -29,22 +29,22 @@ class KernelSumValidator
             [](const UTXO::CPtr& pUTXO) { return pUTXO->GetCommitment(); }
         );
 
-        uint64_t total_mw_supply = 0;
+        uint64_t total_pegins = 0;
+        uint64_t total_pegouts = 0;
         for (const Kernel& kernel : kernels)
         {
             if (kernel.IsPegIn()) {
-                total_mw_supply += kernel.GetAmount();
+                total_pegins += kernel.GetAmount();
```

Comment:
Making this change would mean at some point in the past, there could've been a negative total supply in the MWEB. That's an illegal state, so I don't think I like this change.

What's the reason for making it? Was there a bug in the existing code?

Reply:
The KernelSumValidator::ValidateState test fails intermittently. TxBuilder sorts the kernels and pegouts may now appear before pegins.

Reply:
Maybe we should modify our kernel sorting consensus rules to always include pegins first, then regular kernels, then pegouts. Thoughts?

Reply:
But your first comment implies that the order should be time-based - so that we can ensure the supply was always positive. If the kernels are sorted in an alternative order then we lose that assurance.
Also is there a distinction between the order within blocks and the order within the state?

Reply:
I'm talking about just sorting within a block, not across the whole kernel set. We can sort however we want within the block, as long as after applying a whole block, the MWEB supply is never negative.

This logic here is looping over the entire kernel set, not just a block, so we have to apply them in order and check that it never goes negative. Otherwise, we don't have guarantees that a past block didn't result in a negative supply for a period of time.

Reply:
Ok, so do we agree that the test is at fault here? And the KernelSumValidator::ValidateState logic is correct?

Finally, do we need to add the check kernel.GetFee() > total_mw_supply?

Reply:
> Maybe we should modify our kernel sorting consensus rules to always include pegins first, then regular kernels, then pegouts

Also, this is probably ok to do but only because we want ValidateState to work on individual blocks.

Reply:
> Finally, do we need to add the check kernel.GetFee() > total_mw_supply?

Yes, I missed that case. Please add that check.